plugin/documents.py

```diff
@@ -606,46 +615,65 @@ def render_highlights_on_main_thread() -> None:
-            lambda res: self._on_complete_error(res, resolve))
-
-    def _on_complete_result(self, response: Optional[Union[dict, List]], resolve: ResolveCompletionsFn,
-                            can_resolve_completion_items: bool, session_name: str) -> None:
-        response_items = []  # type: List[Dict]
-        flags = 0
+
+        completion_promises = []  # type: List[Promise[ResolvedCompletions]]
+        for session in sessions:
+
+            def completion_request() -> Promise[ResolvedCompletions]:
+                return session.send_request_task(
+                    Request.complete(text_document_position_params(self.view, location), self.view)
+                ).then(lambda response: (response, session.config.name))
+
+            completion_promises.append(completion_request())
+
+        LspResolveDocsCommand.completions = []
+        Promise.all(completion_promises).then(
+            lambda responses: self._on_all_settled(responses, resolve_completion_list))
+
+    def _on_all_settled(
+        self,
+        responses: List[ResolvedCompletions],
+        resolve_completion_list: ResolveCompletionsFn
+    ) -> None:
+        items = []  # type: List[sublime.CompletionItem]
+        flags = 0  # int
         prefs = userprefs()
         if prefs.inhibit_snippet_completions:
             flags |= sublime.INHIBIT_EXPLICIT_COMPLETIONS
         if prefs.inhibit_word_completions:
             flags |= sublime.INHIBIT_WORD_COMPLETIONS
-        if isinstance(response, dict):
-            response_items = response["items"] or []
-            if response.get("isIncomplete", False):
-                flags |= sublime.DYNAMIC_COMPLETIONS
-        elif isinstance(response, list):
-            response_items = response
-        response_items = sorted(response_items, key=lambda item: item.get("sortText") or item["label"])
-        LspResolveDocsCommand.completions = response_items
-        items = [format_completion(response_item, index, can_resolve_completion_items, session_name)
-                 for index, response_item in enumerate(response_items)]
+
+        for response, session_name in responses:
+            if isinstance(response, Error):
+                sublime.status_message('Completion error: {}'.format(response))
+                continue
+
+            session = self.session_by_name(session_name)
+            if not session:
+                continue
+
+            response_items = []  # type: List[CompletionItem]
+            if isinstance(response, dict):
+                response_items = response["items"] or []
+                if response.get("isIncomplete", False):
+                    flags |= sublime.DYNAMIC_COMPLETIONS
+            elif isinstance(response, list):
+                response_items = response
+            response_items = sorted(response_items, key=lambda item: item.get("sortText") or item["label"])
```

Comment:
Shouldn't this be a global sort, sorting over all completions?

Reply:
The sorting keys of the different language servers are likely to be unrelated to each other. So sorting globally would be a bit odd. I guess how it's written now is okay then.